README.md
# Website-Render
Make your website faster for the user with minimal changes!

**Current version:** V1.0.6. The changes are listed here.
* Fixed support for safari (And other webbrowsers on ios devices)
* Fixed a bug with the "FORCESCRIPT" tag
* Made the loading of webpages a bit faster, every nano second counts!

**Bugs, feature proposal or support:** Click [here](https://github.com/Paultje52/website-render/issues/new/choose).

## Menu
* [Why](#why)
* [How](#how)
* [Docs](#docs)
* [A small example](#example)
* [API](#api)

# Why
You can make very fast websites for users with Website-Render.

What Website-Render does, is load the pages of your website in the background and display them instantly when the user wants to.

Website-Render users express and ejs.

# How?
The back-end code of Website-Render is just like that of an [express](https://expressjs.com) app with [ejs](https://ejs.co/).

When you create your back-end with Website-Render, just use the express syntax (app.get, with req and res, etc). You can use `app.page` for a page that will load on the background.

**Load in the background? What do you mean?**
When a user loads a page, all the pages will be loaded in the background. And when the user clicks on a page, the new page will be displayed right away!

**How?**
All the pages are sent to the user via a websocket with [socket.io](https://socket.io/).

**Is it safe?**
Yes, it's safe. Before the pages are loaded in the background, a call will be made to a test endpoint. And for every page, the test data will be sent, so you can still check the cookies etc.

# Docs
## Docs menu
* [App Constructor](#app-constructor)
* [Express Methods](#express-methods)
* [Set](#set)
* [Page](#page)
* [Start](#start)
* [Front-end](#front-end)

## App Constructor
When you load in Website-Render, you will get a class. You can create your app by constructing.

There is only one parameter while constructing the app: The portnumber.
This needs to be a number AND between 80 and 65535.
```js
const websiteRender = require("website-render");
let app = new websiteRender(8080);
```

## Express methods
The `app` class supports a couple express methods:
* These http methods: `checkout`, `copy`, `delete`, `get`, `options`, `patch`, `post`, `purge`, `put`, `search`, `trace`, `subscribe` and `unsubscribe`. Just use `app.METHOD`
* Enable and Disable: `app.enable(thing)` or `app.disable(thing)`
* Use: `app.use()`
* In the object parameters for the methods and use: `req` and `res` and all the variables/functions with it.

For example, you can make a get and a post method.
```js
const websiteRender = require("website-render");
let app = new websiteRender(8080);
app.get("/", (req, res) => {
  // ...
});
app.post("/", (req, res) => {
  // ...
});
```

## Set
To change settings, you can use the set method: `app.set("key", "value")`.
You can change this.
| Key      | Explanation             | Value type | Example |
| -------- |:---------------------| -------  | ------- |
| Port     | The port that wil be used | Number     | app.set("port", 8080);
| testurl      | The url that will be called for testing permissions    | String | app.set("testurl", "test") |
| frontEndUrl | The url for the front-end code | String | app.set("frontEndUrl", "websiteRender") |

## Page
You can use this just like `app.get`. Every endpoint with `app.page` will be send to the user to load in the background.

Here you get, just like `app.get`, the `req` and `res` parameters. But, there is one difference. You can only use `res.send` and `res.render` with ejs to send pages to the user. Otherwise the page won't be sent to the user.

```js
app.page("/", (req, res) => {
  res.render("index.ejs", {
    ip: req.ip
  });
});
```

## Start
You need to start your app when you loaded the pages, express endpoints, etc. It's very important to start your app as last, because the websocket will be made then.

```js
// Creating an app and making the endpoints
app.start(() => {
  console.log(`Online on ${app.port}!`);
});
```
You can make a function for the start, or you can use await (if you are in an async environment).
```js
// This is an async environment
// Creating an app and making the endpoints
await app.start();
console.log(`Online on ${app.port}!`);
```

## Front-end
### **Background code**
You also need to add one line of code in the head of your front-end html code. You need to put this in every page in the header! This is to load in the front-end code.

```html
<head>
  <script src="/websiteRender.js"></script>
  <!-- Your other stuff -->
</head>
```
By default, the front-end code is located at `/websiteRender.js`, but you can change it with `app.set("frontEndUrl", "URL")`, for example:

Server.js:
```js
// Creating app
app.set("frontEndUrl", "frontEndCode");
// Other stuff
```
All your html pages:
```html
<head>
  <script src="/frontEndCode.js"></script>
  <!-- Your other stuff -->
</head>
```
### **Buttons**
For the button functionality, you need to add this to every button `onclick` event that you want to use to redirect to an other page.
```js
WebsiteRenderButton(`/PATH`)
```
For example, this is a button (with w3css) that redirects to the "about" page.
```html
<button class="w3-button w3-green" onclick="WebsiteRenderButton(`/about`)">About</button>
```
### Loading indication
If you want to show a loading indication until website-render loaded every page, you can do that by setting `window.showLoadingIndication` to `true` before loading the front-end code. Here is an example.
```html
<head>
  <script>
    window.showLoadingIndication = true;
  </script>
  <script src="/websiteRender.js"></script>
  <!-- Your other stuff -->
</head>
```

### Other tags
Website-Render does some thing with two tags: the title tag and the script tag.

**Title tag:**
The title tag is the only thing that will be sent from the head. There is nothing else from the head tag that website-render will sent, so save bandwidth.
**Script tag:**
Website-render will only use a script tag once per page. This is because otherwise there will be some problems because scripts will be executed multiple times.
If you have a script that you want to execute each time a user gets a page, use the tag \<FORCESCRIPT> and \</FORCESCRIPT> in stead of \<script> and \</script>.

**And that's it!**


# Example
You can find an example [here](/test).

# API
## API Menu
* [Class `website-render`](#class-website-render)
  * [Constructor](#constructor)
  * [Function: Set](#function-set)
  * [Function: Page](#function-page)
  * [Function: METHOD](#function-method)
* [Front-end](#front-end-things)
  * [When loaded](#when-loaded)
  * [Function: Loaded](#function-loaded)
  * [Function: Wait](#function-wait)
  * [Function: WebsiteRenderButton](#function-websiterenderbutton)
  * [Function: Remove scripts](#function-removescripts)
## Class `website-render`
### > **Constructor**
Crates an app for `website-render`.

**Parameters:**
* Port: Required: Number between 80 and 65535.

**Returns:**
_Nothing_


### > **Function: Set**
Set's a key to a value in the app.

**Parameters:**
* Key: Required: String
* Value: Required: Anything

**Returns:**
_Nothing_


### > **Function: Page**
Crates a page to load in the background. For normal http requests, this is just an endpoint.

**Parameters:**
* Endpoint: Required: String
* Function to call: Required: Function -> Parameters: [req](https://expressjs.com/en/4x/api.html#req) and [res](https://expressjs.com/en/4x/api.html#res) (From [express](https://expressjs.com/))

**Returns:**
_Nothing_


### > **Function: METHOD** 
METHOD can be: checkout, copy, delete, get, options, patch, post, purge, put, search, trace, subscribe and unsubscribe.

Just normal express endpoints. More information [here](https://expressjs.com/en/4x/api.html#app.METHOD).

## Front-end Things
### > **When loaded**
When the front-end code is loaded, it will load "socketio" and "jquery" automaticly.<br>
When it's loaded in the first page, it will also call the function [loaded](#Function-Loaded).

### > **Function: Loaded**
This function will connect to the websocket and load all the pages.

### > **Function: Wait**
This function is meant for internal use. It needs one parameter, the time to wait in MS.<br>
The only thing it does is promify "setTimeout".

### > **Function: WebsiteRenderButton**
With tis function you can make your buttons.

**Parameters**
* Path: The path for the page you want to make visable, for example `/about`. **Required**
* WaitOnNotFullyRendered: When website-render didn't load all the pages yet, you can choose to let the user wait until all the pages are loaded. When this is true, the user will see a loading icon until website-render is done. **Optional**, default value = true.
* AddToHistory: If website-render needs to add the old page to the history (So the user can click the back button in the browser). **Optional**, default value = true.

**Returns:**
_Nothing_.

### > **Function: RemoveScripts**
This function removes all the \<script> tags with everything within it. It replaces the \<FORCESCRIPT> tags with \<script>.

**Parameters**
* Html: The html code to change. **Required**

**Returns**<br>
The changed html code.

CHANGES.md
# Changes
To see what changed each version, look here.

# Menu
* [V1.0](#v1.0)
  * [V1.0.7](#V1.0.7)
  * [V1.0.6](#V1.0.6)
  * [V1.0.5](#V1.0.5)
  * [V1.0.4](#V1.0.4)
  * [V1.0.3](#V1.0.3)
  * [V1.0.2](#V1.0.2)
  * [V1.0.1](#v1.0.1)
  * [V1.0.0](#v1.0.0)

# V1.0
## V1.0.7
- Fixed package.json
## V1.0.6
* Fixed support for safari (And other webbrowsers on ios devices)
* Fixed a bug with the "FORCESCRIPT" tag
* Made the loading of webpages a bit faster, every nano second counts!
## V1.0.5
* Fixed a bug with the loading icon
* Fixed a bug with the "FORCESCRIPT" tag
* Some minor changes to the documentation.
## V1.0.4
* Website-render won't throw errors anymore
* Website-render only loads scripts once for each page, so you won't get errors from that anymore!
* Website-render will now wait when the button the button is clicked and not all pages are loaded yet
* You no longer need to put the title in the body, if it's in the head, website-render will automaticly detect it and send it to the client.
* Added a way to show a loading icon until website-render loaded the hole page.
* Added comments in [__frontEnd.js](src/__frontEnd.js) (Not the javascript that will be sent to the user), so everyone knows what's going on.
* Some minor changes for the documentation.
## V1.0.3
* Fixed a bug with loading express routers in other files
## V1.0.2
* A bug fix with the versions
## V1.0.1
* Changed something in the documentation
## V1.0.0
* The release with the src, test and documentation


test/index.js
// First: Let's load in the package.
const websiteRender = require("../src/index.js"); // website-render
// Then we will make an app instance on port 8080
let app = new websiteRender(8080);

// We can just use express in our website render app.
app.use((req, _res, next) => {
  console.log(`Request from ${req.ip} to ${req.url}`);
  next();
});

/* 
 Now, let's make some pages! 
 To make pages load in the background of your website, you need to use `app.page`.
 You can use `app.get` or other methods for a normal express app.
*/
app.page("/", (_req, res) => { // Index
  res.render("index.ejs", {
    name: "Test"
  });
});
app.page("/about", (_req, res) => { // About
  res.render("about.ejs", {
    name: "About"
  });
})
app.page("/news", (_req, res) => { // News
  res.render("news.ejs", {
    name: "News"
  });
});

// Last, but not least, let's start the app
app.start(() => { // It's important to start the app as last, because otherwise the `app.use` methods will not be called for the test
  console.log("Online!");
});

test/server.js
// First: Let's load in the package.
const websiteRender = require("website-render");
// Then we will make an app instance on port 8080
let app = new websiteRender(8080);

// We can just use express in our website render app.
app.use((req, _res, next) => {
  console.log(`Request from ${req.ip} to ${req.url}`);
  next();
});

/* 
 Now, let's make some pages! 
 To make pages load in the background of your website, you need to use `app.page`.
 You can use `app.get` or other methods for a normal express app.
*/
app.page("/", (_req, res) => { // Index
  res.render("index.ejs", {
    name: "Test"
  });
});
app.page("/about", (_req, res) => { // About
  res.render("about.ejs", {
    name: "About"
  });
})
app.page("/news", (_req, res) => { // News
  res.render("news.ejs", {
    name: "News"
  });
});

// Last, but not least, let's start the app
app.start(() => { // It's important to start the app as last, because otherwise the `app.use` methods will not be called for the test
  console.log("Online!");
});

src/frontEnd.js
if(void 0===window.pages){if("undefined"==typeof socketio){let t=document.createElement("script");t.src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.3.0/socket.io.js",document.getElementsByTagName("head")[0].appendChild(t)}if("undefined"==typeof jquery){let t=document.createElement("script");t.src="https://code.jquery.com/jquery-3.4.1.min.js",document.getElementsByTagName("head")[0].appendChild(t)}function dockReady(t){let e=setInterval(()=>{try{$(document).ready(()=>{}),io.Manager,clearInterval(e),t()}catch(t){}})}if(dockReady(()=>{loaded(),$("body").empty(),WebsiteRenderButton(`/${window.location.path||""}`,!0,!1,!1)}),window.loading=!1,window.showLoadingIndication){let t=document.createElement("div");t.style='display: block;position: absolute;top: 0;left: 0;z-index: 100;width: 100vw;height: 100vh;background-color: rgba(192, 192, 192, 0.5);background-image: url("https://i.stack.imgur.com/MnyxU.gif");background-repeat: no-repeat;background-position: center;',t.id="websiteRenderLoadingIcon",setTimeout(()=>{document.getElementsByTagName("body")[0].appendChild(t)},100),window.loading=!0}window.pages={},window.pageTitles={};let t=!1;function wait(t){return new Promise(e=>{setTimeout(e,t)})}async function loaded(){let e,o=0,i=0;try{e=io(`http://${window.location.hostname}:${window.location.port}`)}catch(t){let o=document.createElement("script");o.src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.3.0/socket.io.js",document.getElementsByTagName("head")[0].appendChild(o),await wait(1e3),e=io(`http://${window.location.hostname}:${window.location.port}`)}e.emit("pages",!1),e.on("page",e=>{window.pages[e.url]=e.body,window.pageTitles[e.url]=e.title||!1,++i===o&&(console.log("[WEBSITE RENDER] Website fully rendered!"),t=!0,window.loading&&$("#websiteRenderLoadingIcon").remove())}),e.on("pages",t=>{$.ajax(`http://${window.location.hostname}:${window.location.port}/${t.testurl}?id=kjb4lt34uojk3b7&from=website-render`).done(i=>{i=i.id,t.array.forEach(t=>{e.emit("page",{path:t,id:i}),o++})})})}function WebsiteRenderButton(e="/",o=!0,i=!0){if(!window.pages[e]&&!o)return window.location.href=`http://${window.location.hostname}:${window.location.port}${e}`;if(window.pages[e]){window.been.includes(e)&&(window.pages[e]=removeScripts(window.pages[e]));let t=window.pages[e].split("<FORCESCRIPT>").join("<script>").split("</FORCESCRIPT>").join("<\/script>");$("body").html(t),t.includes("<title>")&&t.includes("</title>")&&(document.title=t.split("<title>")[1].split("</title>")[0]),window.pageTitles[e]&&(document.title=window.pageTitles[e]),i&&history.pushState(e,`${document.title} `,`http://${window.location.hostname}:${window.location.port}${e}`),window.been.push(e)}else{let o=document.createElement("div");o.style='display: block;position: absolute;top: 0;left: 0;z-index: 100;width: 100vw;height: 100vh;background-color: rgba(192, 192, 192, 0.5);background-image: url("https://i.stack.imgur.com/MnyxU.gif");background-repeat: no-repeat;background-position: center;',document.title=`Loading ${e}`,$("body").append(o);let n=setInterval(()=>{if(t){if(clearTimeout(d),clearInterval(n),!window.pages[e])return window.location.href=`http://${window.location.hostname}:${window.location.port}${e}`;let t=window.pages[e].split("<FORCESCRIPT>").join("<script>").split("</FORCESCRIPT>").join("<\/script>");$("body").html(t),t.includes("<title>")&&t.includes("</title>")&&(document.title=t.split("<title>")[1].split("</title>")[0]),window.pageTitles[e]&&(document.title=window.pageTitles[e]),i&&history.pushState(e,`${document.title} `,`http://${window.location.hostname}:${window.location.port}${e}`),window.been.push(e)}},250),d=setTimeout(()=>{window.location.href=`http://${window.location.hostname}:${window.location.port}${e}`},2500)}}function removeScripts(t){if(!t.includes("script"))return t;let e="",o=0,i=t=>{t.includes("<script>")&&(0===o&&(e+=t.split("<script>")[0]),o++,i(t.split("<script>").slice(1).join("<script>"))),t.includes("<\/script>")&&(t=t.split("<\/script>")[t.split("<\/script>").length-1],o--),o<=0&&(e+=t)};return i(t),e=e.split("<forcescript>").join("<script>").split("</forcescript>").join("<\/script>")}window.been=[window.location.pathname],window.onpopstate=(()=>{history.state&&WebsiteRenderButton(history.state,!0,!1)})}else rendered=!0;

src/__frontEnd.js
// Where the magic happens
if (window.pages === undefined) {
  // Importing socket.io
  if (typeof socketio === "undefined") {
    let socketio = document.createElement("script");
    socketio.src = "https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.3.0/socket.io.js";
    document.getElementsByTagName("head")[0].appendChild(socketio);
  }
  // Importing jquery
  if (typeof jquery === "undefined") {
    let jquery = document.createElement("script");
    jquery.src = "https://code.jquery.com/jquery-3.4.1.min.js";
    document.getElementsByTagName("head")[0].appendChild(jquery);
  }

  // Checking if the document is ready (Checking if you can use socket.io and jquery)
  function dockReady(fn) {
    let i = setInterval(() => {
      try {
        $(document).ready(() => { });
        io.Manager;
        clearInterval(i);
        fn();
      } catch (e) { }
    });
  }

  // When the document is ready
  dockReady(() => {
    loaded();
    $("body").empty();
    WebsiteRenderButton(`/${window.location.path || ""}`, true, false, false);
  });

  // Show the loading icon until website-render loaded everything.
  window.loading = false;
  if (window.showLoadingIndication) {
    let loadingElement = document.createElement("div");
    loadingElement.style = `display: block;position: absolute;top: 0;left: 0;z-index: 100;width: 100vw;height: 100vh;background-color: rgba(192, 192, 192, 0.5);background-image: url("https://i.stack.imgur.com/MnyxU.gif");background-repeat: no-repeat;background-position: center;`;
    loadingElement.id = "websiteRenderLoadingIcon";
    setTimeout(() => {
      document.getElementsByTagName("body")[0].appendChild(loadingElement);
    }, 100);
    window.loading = true;
  }
  // Presetting
  window.pages = {};
  window.pageTitles = {};
  let rendered = false;

  function wait(time) {
    return new Promise((res) => {
      setTimeout(res, time);
    });
  }

  // On page load
  async function loaded() {
    let waiting = 0;
    let i = 0;
    // Starting the websocket
    let socket;
    try {
      socket = io(`http://${window.location.hostname}:${window.location.port}`);
    } catch (e) {
      let socketio = document.createElement("script");
      socketio.src = "https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.3.0/socket.io.js";
      document.getElementsByTagName("head")[0].appendChild(socketio);
      await wait(1000);
      socket = io(`http://${window.location.hostname}:${window.location.port}`);
    }
    // Asking for the website pages
    socket.emit("pages", false);
    // For each page
    socket.on("page", (data) => {
      window.pages[data.url] = data.body;
      window.pageTitles[data.url] = data.title || false;
      i++;
      if (i === waiting) {
        console.log("[WEBSITE RENDER] Website fully rendered!");
        rendered = true;
        if (window.loading) $("#websiteRenderLoadingIcon").remove();
      }
    });
    // When getting all the pages
    socket.on("pages", (object) => {
      // Sending the test url
      $.ajax(`http://${window.location.hostname}:${window.location.port}/${object.testurl}?id=kjb4lt34uojk3b7&from=website-render`).done((id) => {
        id = id.id;
        object.array.forEach(a => {
          // Sending the individual page requests
          socket.emit("page", { path: a, id: id });
          waiting++;
        });
      });
    });
  }

  // Website Render Button Function
  window.been = [window.location.pathname];
  function WebsiteRenderButton(path = "/", waitOnNotFullyRendered = true, addHistory = true) {
    // If the page isn't loaded and website-render isn't done yet
    if (!window.pages[path] && !waitOnNotFullyRendered) return window.location.href = `http://${window.location.hostname}:${window.location.port}${path}`;
    if (window.pages[path]) { // When the requested page is loaded
      if (window.been.includes(path)) window.pages[path] = removeScripts(window.pages[path]);
      let toSend = window.pages[path].split("<FORCESCRIPT>").join("<script>").split("</FORCESCRIPT>").join("</script>");
      $("body").html(toSend);
      if (toSend.includes("<title>") && toSend.includes("</title>")) document.title = toSend.split("<title>")[1].split("</title>")[0];
      if (window.pageTitles[path]) document.title = window.pageTitles[path];
      if (addHistory) history.pushState(path, `${document.title} `, `http://${window.location.hostname}:${window.location.port}${path}`);
      window.been.push(path);
    } else { // When the page isn't loaded yet
      // Loading indication
      let loading = document.createElement("div");
      loading.style = `display: block;position: absolute;top: 0;left: 0;z-index: 100;width: 100vw;height: 100vh;background-color: rgba(192, 192, 192, 0.5);background-image: url("https://i.stack.imgur.com/MnyxU.gif");background-repeat: no-repeat;background-position: center;`;
      document.title = `Loading ${path}`;
      $("body").append(loading);
      // Interval for checking if website-render is done
      let int = setInterval(() => {
        if (rendered) { // When website-render is done
          clearTimeout(timeout);
          clearInterval(int);
          if (!window.pages[path]) return window.location.href = `http://${window.location.hostname}:${window.location.port}${path}`;
          let toSend = window.pages[path].split("<FORCESCRIPT>").join("<script>").split("</FORCESCRIPT>").join("</script>");
          $("body").html(toSend);
          if (toSend.includes("<title>") && toSend.includes("</title>")) document.title = toSend.split("<title>")[1].split("</title>")[0];
          if (window.pageTitles[path]) document.title = window.pageTitles[path];
          if (addHistory) history.pushState(path, `${document.title} `, `http://${window.location.hostname}:${window.location.port}${path}`);
          window.been.push(path);
        }
      }, 250);
      // Fail safe, for if something went wrong while rendering (Like an server error)
      let timeout = setTimeout(() => {
        // 2,5 seconds after clicking the button, the page will be force-reloaded
        window.location.href = `http://${window.location.hostname}:${window.location.port}${path}`;
      }, 2500);
    }
  }

  // Function to remove the scripts for every page that allready loaded before.
  function removeScripts(html) {
    if (!html.includes("script")) return html;
    let output = "";
    let level = 0;
    let handle = (s) => {
      if (s.includes("<script>")) {
        if (level === 0) output += s.split("<script>")[0];
        level++;
        handle(s.split("<script>").slice(1).join("<script>"));
      }
      if (s.includes("</script>")) {
        s = s.split("</script>")[s.split("</script>").length - 1];
        level--;
      }
      if (level <= 0) output += s;
    }
    handle(html);
    output = output.split("<forcescript>").join("<script>").split("</forcescript>").join("</script>");
    return output;
  }

  // For the "back" button in browsers
  window.onpopstate = () => {
    if (history.state) WebsiteRenderButton(history.state, true, false);
  }
} else rendered = true;

src/index.js
const express = require("express");
const http = require("http");
const io = require("socket.io");
const randomString = require("random-string");
const ejs = require("ejs");
const fs = require("fs");

module.exports = exports = class WebsiteRender {
  constructor(port) {
    if (!port || typeof port !== "number" || port < 80 || port > 65535) throw new error("The listen port needs to be as number, and between 80 and 65535.");
    this.port = port;
    this.testurl = "test";
    this.express = express();
    this.pages = {};
    this.frontEndUrl = "websiteRender";
    let methods = ["enable", "disable", "use", "engine", "checkout", "copy", "delete", "get", "options", "patch", "post", "purge", "put", "search", "trace", "subscribe", "unsubscribe"];
    methods.forEach((method) => {
      this[method] = (path, callback = false) => {
        try {
          callback.stack[0];
          this.express[method](path, callback);
        } catch (e) {
          if (!["use", "enable", "disable"].includes(method)) this.express[method](path, callback);
          else this.express[method](path);
        }
      }
    });
  }

  set(key, value) {
    this[key] = value;
  }

  page(path, callback) {
    this.pages[path] = callback;
  }

  async start(func = () => { }) {
    return new Promise((res) => {
      this.tests = {};
      this.express.get(`/${this.testurl}`, (req, res, next) => {
        if (!req.query.id) return next();
        if (!req.query.from || req.query.from !== "website-render") return next();
        let id = req.query.id + randomString({ length: 15 });
        this.tests[id] = { req: req, res: res };
        res.json({ message: "Done!", id: id });
      });
      for (let i in this.pages) {
        // i = path
        // this.pages[i] = callback
        let func = async (req, res, next) => {
          if (this.checkRun) {
            let result = await this.checkRun(req, res);
            if (!result) return;
          }
          req.direct = false;
          res.doNotLoad = () => { res.json({ status: 406, loadInBackground: false }) };
          res.render = (file, information = {}) => {
            file = `${process.cwd()}/views/${file}`;
            file = fs.readFileSync(file).toString("utf8");
            file = ejs.render(file, information);
            // file = file.split("<FORCESCRIPT>").join("<script>").split("</FORCESCRIPT>").join("</script>");
            res.send(file);
          }
          this.pages[i](req, res, next);
        }
        this.express.get(i, func);
      }
      this.express.get(`/${this.frontEndUrl}.js`, (_req, res) => {
        res.sendFile(`${__dirname}/__frontEnd.js`);
      });

      this.http = http.createServer(this.express);
      this.io = io(this.http);
      this.http.listen(this.port, () => {
        func();
        res();
      });

      this.io.on("connection", (socket) => {
        socket.on("pages", (check) => {
          if (!check) {
            let array = [];
            for (let i in this.pages) {
              array.push(i);
            }
            socket.emit("pages", {
              array: array,
              testurl: this.testurl
            });
          }
        });

        socket.on("page", async (information) => {
          if (!this.tests[information.id]) return;
          let testData = this.tests[information.id];
          let result = () => {
            return new Promise((_mkwe_24jfds_resolve) => {
              testData.res.send = (data) => {
                _mkwe_24jfds_resolve(data);
              }
              testData.res.render = (file, information = {}) => {
                file = `${process.cwd()}/views/${file}`;
                file = fs.readFileSync(file).toString("utf8");
                file = ejs.render(file, information);
                _mkwe_24jfds_resolve(file);
              }
              this.pages[information.path](testData.req, testData.res);
            });
          }
          result = await result();
          let title = false;
          if (result.includes("<title>")) title = result.split("<title>")[1].split("</title>")[0];
          if ((result.includes("<body>") || result.includes("<body")) && result.includes("</body>")) {
            result = result.split("<body>")[1] || result.split("<body")[1].split(">").slice(1).join(">");
            result = result.split("</body>")[0];
          }
          socket.emit("page", {
            url: information.path,
            body: result,
            title: title
          });
        });
      });
    });
  }
}

exports.version = "v1.0.7";
